server/text_generation_server/models/model.py

```diff
@@ -33,11 +24,26 @@ def batch_type(self) -> Type[B]:
     def generate_token(self, batch: B) -> Tuple[List[GeneratedText], Optional[B]]:
         raise NotImplementedError
 
-    def decode_token(self, token_id: int) -> str:
+    def decode_token(
+        self, all_input_ids: List[int], offset: Optional[int] = None
+    ) -> Tuple[str, Optional[int]]:
         """Hack to hopefully support generate_stream for the maximum number of tokenizers"""
-        # append token to special decode token and decode both
-        result = self.tokenizer.decode(
-            [self.special_decode_token_id, token_id], skip_special_tokens=False
+
+        # Decode all token minus last one and all tokens
+        results = self.tokenizer.batch_decode(
+            [all_input_ids[:-1], all_input_ids],
```

Comment:
For large outputs, might this affect performance (repeatedly decoding all tokens)?

Reply:
Yes.
This branch should not be merged in this state.

Reply:
@OlivierDehaene what do you think about the idea of encapsulating this logic into an incremental decoder as I was proposing [here](https://github.com/huggingface/text-generation-inference/blob/65ff6a73b3d5a3bb0cafa3d667b0d077bfcfc3f5/router/src/decoder.rs#L284-L291) (in rust but I guess similar could be done in python). Which can handle all of the tokenizer-specific aspects and unicode boundaries etc. Just feed it the token ids and take back strings to be concatenated. This impl has been working very well for us.

Reply:
By the way this also handles the subtle difference between the decoder and encoder-decoder cases, where for decoder-only it's preferable to just start concatenating to the pre-tokenized input text rather than roundtripping it.